### PR TITLE
Implement an `extractall` function

### PR DESCRIPTION
**Pull request by DragonMinded**
## Summary

This is equivalent to calling `findall` on a node, and then calling `extract` with the same
matcher on each of the returned nodes from `findall`.

## Test Plan

tox, added tests. This is a refactor of `findall` so I expect most edge cases to be covered already. I only tested the basic cases as a result.

### CODE REVIEW

File: libcst/matchers/tests/test_findall.py [hunk 1]
```diff
@@ -162,3 +162,34 @@ def foo(bar: int) -> bool:
...
+            raise AssertionError(
+                f"\n{seq1!r}\nis not deeply equal to \n{seq2!r}{suffix}"
+            )
+
+        for node1, node2 in zip(seq1, seq2):
+            if not node1.deep_equals(node2):
+                raise AssertionError(
+                    f"\n{seq1!r}\nis not deeply equal to \n{seq2!r}{suffix}"
+                )
+
```
Comment (jimmylai) [inline libcst/matchers/tests/test_findall.py]: Can we have some tests to cover this part?

**Comment by jimmylai** [comment]
Trying to check the generated doc  but for some reason, the uploaded artifacts tab didn't show up in circleci job even though the artifacts seemed uploaded successfully. https://circleci.com/gh/Instagram/LibCST/3991#tests/containers/0

File: libcst/matchers/tests/test_findall.py (hunk 1 above)
Comment (DragonMinded) [reply libcst/matchers/tests/test_findall.py]: You want tests that cover the tests?

Comment (jimmylai) [reply libcst/matchers/tests/test_findall.py]: I know it's test helper, just feel like we probably don't want to add code that don't run but this can be used to guard future test cases.
This code block is pretty much the same as MatchersFindAllTest, can we just reuse it? E.g. keep those matcher test helpers in a base class.

https://github.com/Instagram/LibCST/blob/675f8b3e8297b362f4de9b1ba6354e35c0fb049f/libcst/matchers/tests/test_findall.py#L25-L35

Comment (DragonMinded) [reply libcst/matchers/tests/test_findall.py]: Its unit test code though. Its not used in production, and if a new test is written that fails, it hits this code. I see no value in testing tests nor trying to commonize this code. Its a waste of engineering effort for no benefit.

Comment (DragonMinded) [reply libcst/matchers/tests/test_findall.py]: Jimmy, this is dead code. I just checked and its not even called. Why are you focusing on test coverage to the detriment of an actual review?

Comment (jimmylai) [reply libcst/matchers/tests/test_findall.py]: In general, duplicated code adds tech debt and maintenance burden for future. I didn't specific focus on test coverage. Code quality is as important as implemented features to me and I review both of them.

Comment (DragonMinded) [reply libcst/matchers/tests/test_findall.py]: You're completely missing the point. I checked the code and it was entirely unused and I deleted it, it looks like a copy-paste mistake. You're focusing on the minutia of "We should have tested code" instead of the bigger picture. This isn't valuable feedback at all.

**Comment by jimmylai** [comment]
image

image

image


The type annotations looks complex and may not be useful to users. We should try to simplify it.

Otherwise LGTM.

**Comment by DragonMinded** [comment]
In general the sphynx generated type hints are way overcomplicated since they unroll everything :(

It is a lot simpler in code: https://github.com/Instagram/LibCST/pull/154/files#diff-4f19c4e1c6cef2912261ee532dff5807R1582-R1593

I think maybe we can get rid of the InverseOf part, since that is hidden from the type system.

**Comment by DragonMinded** [comment]
I got rid of the InverseOf part and it halved the length of the type hints so I think that is a pretty good improvement. I still want to eventually tackle printing the exported name instead of the full name (so `libcst.matchers.MatchIfTrue` instead of `libcst.matchers._matcher_base.MatchIfTrue` but this is still an improvement.
